pkg/ingestor/collector/gcs/gcs.go

```diff
@@ -0,0 +1,163 @@
+	bucket       string
+	reader       gcsReader
+	lastDownload time.Time
+	isDone       bool
+}
+
+const (
+	// Specify the GCS bucket address
+	bucketEnv = "GCS_BUCKET_ADDRESS"
+)
+
+func init() {
+	collector.RegisterDocumentCollector(&GCS{}, collector.CollectorGCS)
+}
+
+func getBucketPath() string {
+	if env := os.Getenv(bucketEnv); env != "" {
+		return env
+	}
+	return ""
+}
+
+// Enabled initializes GCS and returns true if properly configured
+func (g *GCS) Enabled(ctx context.Context) bool {
+	if getBucketPath() != "" {
+		client, err := storage.NewClient(ctx)
+		if err != nil {
+			return false
+		}
+		g.bucket = getBucketPath()
+		g.reader = &reader{client: client, bucket: getBucketPath()}
+		g.isDone = true
+		return true
+	}
+	return false
+}
+
+// IsDone return if the collector is done collecting artifacts
+func (g *GCS) IsDone() bool {
+	return g.isDone
+}
+
+// Type is the collector type of the collector
+func (g *GCS) Type() collector.CollectorType {
+	return collector.CollectorGCS
+}
+
+type gcsReader interface {
+	getIterator(ctx context.Context) *storage.ObjectIterator
+	getReader(ctx context.Context, object string) (io.ReadCloser, error)
+}
+
+type reader struct {
+	client *storage.Client
+	bucket string
+}
+
+func (r *reader) getIterator(ctx context.Context) *storage.ObjectIterator {
+	q := &storage.Query{
+		Projection: storage.ProjectionNoACL,
+	}
+	// set query to return only the Name and Updated attributes
+	q.SetAttrSelection([]string{"Name", "Updated"})
+	return r.client.Bucket(r.bucket).Objects(ctx, q)
+}
+
+func (r *reader) getReader(ctx context.Context, object string) (io.ReadCloser, error) {
+	return r.client.Bucket(r.bucket).Object(object).NewReader(ctx)
+}
+
+// RetrieveArtifacts get the artifacts from the collector source
+func (g *GCS) RetrieveArtifacts(ctx context.Context) ([]*processor.Document, error) {
+
+	artifacts := []*processor.Document{}
+	if g.reader == nil {
+		return nil, errors.New("GCS reader not initialized")
+	}
+	it := g.reader.getIterator(ctx)
+	g.isDone = false
+	for {
+		attrs, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, fmt.Errorf("failed to retrieve object attribute from bucket: %s", g.bucket)
+		}
+		payload := []byte{}
+		if g.lastDownload.IsZero() {
+			payload, err = g.getObject(ctx, attrs.Name)
+			if err != nil {
+				logrus.Warnf("failed to retrieve object: %s from bucket: %s", attrs.Name, g.bucket)
+				continue
+			}
+		} else if attrs.Updated.After(g.lastDownload) {
+			payload, err = g.getObject(ctx, attrs.Name)
+			if err != nil {
+				logrus.Warnf("failed to retrieve object: %s from bucket: %s", attrs.Name, g.bucket)
+				continue
+			}
+		}
+		if len(payload) > 0 {
+			artifacts = append(artifacts, &processor.Document{
+				Blob: payload,
+				SourceInformation: processor.SourceInformation{
+					Collector: collector.CollectorGCS,
+					Source:    g.bucket,
+				},
+			})
+		}
+	}
+	g.lastDownload = time.Now()
+	g.isDone = true
```

Comment:
hmm would that be a mode that this would keep running to find new docs?

Reply:
done.

Reply:
added poller